src/devices/src/virtio/block/device.rs

```diff
@@ -0,0 +1,263 @@
+use crate::virtio::{
+    MmioConfig, SingleFdSignalQueue, QUEUE_MAX_SIZE, VIRTIO_MMIO_QUEUE_NOTIFY_OFFSET,
+};
+
+use super::inorder_handler::InOrderQueueHandler;
+use super::queue_handler::QueueHandler;
+use super::{build_config_space, BlockArgs, Error, Result};
+
+// This Block device can only use the MMIO transport for now, but we plan to reuse large parts of
+// the functionality when we implement virtio PCI as well, for example by having a base generic
+// type, and then separate concrete instantiations for `MmioConfig` and `PciConfig`.
+pub struct Block<M: GuestAddressSpace> {
+    virtio_cfg: VirtioConfig<M>,
+    mmio_cfg: MmioConfig,
+    endpoint: RemoteEndpoint<Arc<Mutex<dyn MutEventSubscriber + Send>>>,
+    vm_fd: Arc<VmFd>,
+    irqfd: Arc<EventFd>,
+    file_path: PathBuf,
+}
+
+impl<M> Block<M>
+where
+    M: GuestAddressSpace + Clone + Send + 'static,
+{
+    pub fn new<B>(mut args: BlockArgs<M, B>) -> Result<Arc<Mutex<Self>>>
+    where
+        // We're using this (more convoluted) bound so we can pass both references and smart
+        // pointers such as mutex guards here.
+        B: DerefMut,
+        B::Target: MmioManager<D = Arc<dyn DeviceMmio + Send + Sync>>,
+    {
+        // The queue handling logic for this device uses the buffers in order, so we enable the
+        // corresponding feature as well.
+
+        // TODO: We don't offer flush (1 << VIRTIO_BLK_F_FLUSH), so writes are faster overall but
+        // not guaranteed to be persisted to physical storage in the event of something like power
+        // loss. Add this as a configuration option?
+        let device_features =
+            1 << VIRTIO_F_VERSION_1 | 1 << VIRTIO_F_IN_ORDER | 1 << VIRTIO_F_RING_EVENT_IDX;
+
+        // A block device has a single queue.
+        let queues = vec![Queue::new(args.common.mem, QUEUE_MAX_SIZE)];
+        let config_space = build_config_space(&args.file_path)?;
+        let virtio_cfg = VirtioConfig::new(device_features, queues, config_space);
+
+        // Used to send notifications to the driver.
+        let irqfd = EventFd::new(EFD_NONBLOCK).map_err(Error::EventFd)?;
+
+        args.common
+            .vm_fd
+            .register_irqfd(&irqfd, args.common.mmio_cfg.gsi)
+            .map_err(Error::RegisterIrqfd)?;
+
+        let mmio_cfg = args.common.mmio_cfg;
+
+        let block = Arc::new(Mutex::new(Block {
+            virtio_cfg,
+            mmio_cfg,
+            endpoint: args.common.event_mgr.remote_endpoint(),
+            vm_fd: args.common.vm_fd,
+            irqfd: Arc::new(irqfd),
+            file_path: args.file_path,
+        }));
+
+        // Register the device on the MMIO bus.
+        args.common
+            .mmio_mgr
+            .register_mmio(mmio_cfg.range, block.clone())
+            .map_err(Error::Bus)?;
+
+        // TODO: We're hardcoding the MMIO range len to 4K so assert for this until we decide
```

Comment:
Since it's hardcoded, it would be better to declare it `const` somewhere.

Reply:
This was a sort-of lazy solution to the problem of converting the length of the MMIO region to a proper textual representation (altho, in retrospect, just using the raw number of bytes should work as well). Looks like this will be nicely fixed by using the addition proposed in https://github.com/rust-vmm/linux-loader/pull/53, and passing a `Cmdline` mutable handle instead of that `String`.